# **Section 01 : Introduction to Nested For Loops**
* This section provides an in-depth exploration of nested for loops in Python, including practical examples, challenges, how these loops can impact performance, and best practices for improving code readability and efficiency.

## **A short recap of basic For Loops**
A **For loop** in Python is used to iterate over a Collection (like a `list`, `tuple`, `string`, or `range`) and perform a specific operation for each element in that collection.

* Here is a simple example to count the number of .txt files from a list of filenames by using For loops:

In [ ]:
files = ["report.txt", "data.csv", "image.png", "notes.txt", "summary.docx"]
count = 0
for file in files:
    if '.txt' in file:                # membership operation on a string
      count +=1
print("Count of text files :",count)

Count of text files : 2


* Here is the List Comprehension way to achieve the same objective

In [ ]:
files = ["report.txt", "data.csv", "image.png", "notes.txt", "summary.docx"]
count = sum([1 for file in files if '.txt' in file])
print("Count of text files :",count)

Count of text files : 2


## **The Need for Nested For Loops**

* Sometimes, you may need to perform more complex operations that require looping through multiple Collections. This is where Nested For loops come useful. A Nested For loop is simply a For loop inside another For loop.

* For instance, if you have a list of lists (a list where each element is another list), and you want to access each item inside these inner lists, you can use a Nested For loop.

##### **Example 1: Printing Elements of a Matrix**
* In this example:
 * The outer loop iterates through each list (row) in the matrix.
 * The inner loop iterates through each item in the current row and prints it.

In [ ]:
matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

for row in matrix:           # Outer loop
    for item in row:         # Inner loop
        print(item, end=" ")
    print()

1 2 3 
4 5 6 
7 8 9 


##### **Example 2: Generating Pairs of Coordinates**
* Nested For loops are also useful for generating pairs of elements, such as coordinates.

In [ ]:
for x in range(3):           # Outer loop for x-coordinate
    for y in range(3):       # Inner loop for y-coordinate
        print(f"({x}, {y})")

(0, 0)
(0, 1)
(0, 2)
(1, 0)
(1, 1)
(1, 2)
(2, 0)
(2, 1)
(2, 2)


## **Challenges with Applying Multiple Levels of For Loops**
While Nested For loops can be powerful, they also come with some challenges:

### **Code Readability**
* The more levels of nesting you have, the more complex and harder to read your code becomes.
* For example, if you're processing a 3D array (a list of lists of lists), the logic can get difficult to follow.
* As the levels of nesting increase, the code can become even more difficult to debug and maintain.

In [ ]:
array_3d = [[[1, 2], [3, 4]],
            [[5, 6], [7, 8]],
            [[9, 10], [11, 12]]]

for layer in array_3d:            # First level
    for row in layer:             # Second level
        for item in row:          # Third level
            print(item, end=" ")
        print()
    print()

1 2 
3 4 

5 6 
7 8 

9 10 
11 12 



### **Performance**
* Multiple nested loops can significantly slow down your program, especially if you're dealing with large datasets.

* **Time Complexity**
  * Time complexity helps us understand how the runtime of an operation (e.g.sorting algorithm) grows as the size of the input data increases.

* **Nested Loops Time Complexity:**

 * When you have nested for loops, the time complexity depends on how many loops you have and what they are doing.
 * The time it takes to run nested loops grows quickly. If you have two nested loops each running n times, the time complexity is O(n^2). Three nested loops result in O(n^3), and so on.

* **Example - Creating a 2D list (matrix) where each element is initialized to 0.**

In [ ]:
n = 5
matrix = []
for i in range(n):
    row = []
    for j in range(n):
        row.append(0)   # Creating a new item in the row
    matrix.append(row)  # Creating a new row in the matrix
matrix

[[0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0]]

* This operation is O(n^2), where n is the number of rows or columns. As n grows, the number of operations grows quadratically, making the loop computationally expensive for large matrices.

* **Example - Creating a 3D list (matrix) where each element is initialized to 0.**

In [ ]:
n = 3
matrix_3d = []
for i in range(n):
    layer = []
    for j in range(n):
        row = []
        for k in range(n):
            row.append(0)
        layer.append(row)
    matrix_3d.append(layer)

matrix_3d

[[[0, 0, 0], [0, 0, 0], [0, 0, 0]],
 [[0, 0, 0], [0, 0, 0], [0, 0, 0]],
 [[0, 0, 0], [0, 0, 0], [0, 0, 0]]]

* This operation is O(n^3), where n is the number of layers, rows, or columns. As n grows, the number of operations grows cubically, making the loop computationally expensive for large 3D matrices.

### **Logical Errors**
* It's easy to introduce logical errors when using Nested For loops.
* For example, accidentally placing an operation in the wrong loop can produce unexpected results.


In [ ]:
matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]

for row in matrix:               # Outer loop
    for item in matrix:          # Inner loop (wrong loop used)
        print(item, end=" ")
    print()

[1, 2, 3] [4, 5, 6] [7, 8, 9] 
[1, 2, 3] [4, 5, 6] [7, 8, 9] 
[1, 2, 3] [4, 5, 6] [7, 8, 9] 


##### **Explanation:**

* The inner loop should iterate through row instead of matrix. This mistake results in incorrect output because the inner loop runs over the entire matrix instead of the current row.

##### **Corrected Code:**

In [ ]:
for row in matrix:               # Outer loop
    for item in row:             # Inner loop (correct loop used)
        print(item, end=" ")
    print()

1 2 3 
4 5 6 
7 8 9 


## **Best Practices to Avoid Nested For Loops**
* Given the potential challenges with Nested For loops, it’s often best to look for alternative approaches, especially when dealing with large datasets:

 * List Comprehensions: In some cases, list comprehensions can simplify writing of Nested For loops, making the code more concise and readable.

 * Vectorization: Use libraries like `NumPy`, which operate on entire arrays or matrices at once.

In [ ]:
list1 = [1, 2, 3]
list2 = ['a', 'b', 'c']

cartesian_product = [(x, y) for x in list1 for y in list2]
print(cartesian_product)

[(1, 'a'), (1, 'b'), (1, 'c'), (2, 'a'), (2, 'b'), (2, 'c'), (3, 'a'), (3, 'b'), (3, 'c')]


* Using Built-in Functions: Functions like `map()`, `filter()`, and `reduce()` can sometimes replace loops, making your code more efficient.

* Using `itertools`: The `itertools` module provides powerful tools to handle iterators, which can sometimes eliminate the need for nested loops.

# **Section 02 : Using `break` and `continue` in For Loops**

### **The Need for break and continue Statements**

In programming, `break` and `continue` are useful for managing how loops execute. They can help make your code more efficient and responsive to specific conditions that might arise during execution.

### **Simple examples to understand their working**

##### **Example of `break` in a For Loop**
* **break Statement:** Useful when you want to stop a loop as soon as a particular condition is met. This is common in scenarios like searching for an item or processing input until a stop condition is encountered.

In [ ]:
for number in range(1, 11):
    if number == 5:
        break  # Skip the rest of the loop for this iteration
    print(number)

1
2
3
4


##### **Example of `continue` in a For Loop**
* **Continue Statement:** Useful when you want to skip certain iterations based on a condition, such as skipping invalid data entries or avoiding unnecessary computations

In [ ]:
for number in range(1, 11):
    if number == 5 :
        continue  # Skip the rest of the loop for this iteration
    print(number)

1
2
3
4
6
7
8
9
10


### **More Practical examples to understand the application of `break` and `continue`**

##### **Example 1: Incrementally Adding File Sizes and Stopping When a Limit is Reached**
* Imagine you're processing a list of file sizes and want to keep adding their sizes until you reach a maximum storage limit. Once the limit is reached, you stop processing further files.

In [ ]:
file_sizes = [150, 200, 75, 300, 50]
storage_limit = 400
current_storage = 0

for size in file_sizes:
    if current_storage + size > storage_limit:
        print(f"Adding file of size {size} MB would exceed the limit. Stopping further processing.")
        break  # Stop further processing as adding this file exceeds the limit
    else:
      current_storage += size
      print(f"Added file of size {size} MB. Current storage used: {current_storage} MB")

print(f"Total storage used: {current_storage} MB")

Added file of size 150 MB. Current storage used: 150 MB
Added file of size 200 MB. Current storage used: 350 MB
Adding file of size 75 MB would exceed the limit. Stopping further processing.
Total storage used: 350 MB


**Explanation**
* Here you are checking each file one by one. First, you try to add a 150 MB file. It fits within the limit, so you add it and update your current storage to 150 MB.
* Next, you add a 200 MB file. It also fits, so now your storage usage is 350 MB.
* When you reach the 75 MB file, you see that adding it would exceed your 400 MB storage limit, bringing the total to 425 MB.
* **At this point, the `break` statement is crucial:** it immediately stops the loop from checking any more files, preventing further additions.
* You print a message to inform that adding this file would surpass the limit, and then halt any further processing.
* After stopping, you print the total storage used, which is 350 MB.

##### **Example 2: Assigning Grades and Handling Non-Passing Scores**
* Consider a scenario where student scores are checked: non-passing scores get a message, while passing scores are further evaluated to assign grades.

In [ ]:
scores = [78, 45, 62, 89, 33, 55, 47]

for score in scores:
  if score < 50:
    print("Did not clear. Re-attempt")
    continue
  elif score <= 60:
    print("Grade D")
  elif score <= 70:
    print("Grade C")
  elif score <= 80:
    print("Grade B")
  elif score <= 90:
    print("Grade A")
  else:
    print("Grade A+")

Grade B
Did not clear. Re-attempt
Grade C
Grade A
Did not clear. Re-attempt
Grade D
Did not clear. Re-attempt


* **Explanation**
 * The for loop iterates through each score in the scores list.
 * If the score is less than 50, it prints "Did not clear. Re-attempt" and uses `continue` to skip to the next iteration (i.e., next score), effectively bypassing the grade assignment for these scores.
 * The `continue` statement ensures that once a score is identified as below 50, the loop skips to the next score without further processing.
 * For scores of 50 or above, the elif statements are used to determine the grade based on the score:

### **Using break in Nested Loops**
* If you are working with nested loops and need to exit only the inner loop, `break` will only terminate the loop it is called in.

* **Example - "Early Target Achievement for Store Sales"**
 * Here we calculate how many weeks it took for each store to meet a Target of 5000 units in sales
 * The program prints the number of weeks it took for each store to reach or exceed 5000 units in total sales.
 * If the target is achieved before processing all 8 weeks, the loop breaks early.

In [ ]:
sales_data = [
    [900, 1000, 900, 1100, 1050, 900, 700, 1150],   # Store 1
    [900, 1200, 100, 900, 800, 900, 1000, 900],     # Store 2
    [1100, 1100, 1000, 900, 1200, 1200, 900, 1000], # Store 3
    [4000, 900, 900, 1100, 1000, 1010, 900, 900],    # Store 4
]

for store,store_sales in enumerate(sales_data):
  tot_sales = 0                   # initializing total sales variable for each store
  count = 0                       # initializing count of weeks variable for each store
  for week_sales in store_sales:
    if tot_sales > 5000:
      break
    else:
      tot_sales += week_sales
      count +=1
  print("Store",store+1,"achieved target in ",count,"weeks")


Store 1 achieved target in  6 weeks
Store 2 achieved target in  7 weeks
Store 3 achieved target in  5 weeks
Store 4 achieved target in  3 weeks


**Explanation:**
* Each item in the list `sales_data` is another list representing the weekly sales numbers for a store over eight weeks.

* We use the `enumerate` function to loop through `sales_data`, which gives us both the index (`store`) and the values (`store_sales`) for each of the inner lists. Note that the index `store` starts from 0.

* For each store, we initialize two variables: `tot_sales` to keep track of the total sales so far, and `count` to count the number of weeks.

* We loop through the weekly sales in each store list. For each week's sales, we add the sales number to `tot_sales`.

* If `tot_sales` exceeds 5000, we break out of the loop for that store. This means we stop counting further weeks once the store reaches the target.

* After each week's sales are added to `tot_sales`, we increment count to track how many weeks it took to reach or exceed 5000 units.

* Finally, we print out how many weeks it took for each store to reach the target. If a store doesn’t reach 5000 units within 8 weeks, the loop will still show the total number of weeks counted.

# **Section 03 : Running For Loop on Two or More Collections/Variables**

## **Background and Need**

* We've used simple `for` loops to iterate through a single collection, such as a List or a Tuple. This has allowed us to process each element in the collection individually.
* We've briefly used the `enumerate()` function to get both the index and the value while iterating through a single collection.

* However, there are scenarios where you might need to iterate over two or more collections simultaneously. For example:

 - **Parallel Data Processing:** When you have multiple related lists or tuples, such as names and scores, and you want to process each pair together.
 - **Combining Information:** When you need to work with multiple datasets that correspond to each other, like combining lists of dates and temperatures.
 - **Synchronizing Iterations:** When the collections have the same length and you want to perform operations on corresponding elements.

* In these situations, the `zip()` function is incredibly useful. It allows you to iterate over multiple sequences at the same time, pairing corresponding elements together in each iteration.


## **Understanding the `zip()` Function**

The `zip()` function combines two or more collections (such as lists or tuples) into tuples. Each tuple contains elements from the iterables at the same position.

### **Zipped object**
* The output `<zip at ...........>` that you see when you run the code simply represents the memory address where the zip object is stored.
* The zip object itself is an iterator, meaning it produces its items one at a time when you iterate over it.
* To view the actual contents of the zip object (the pairs of names and scores), you need to either run a For loop on it or convert it to a list or another iterable type (tuple, etc)

In [ ]:
names = ["Alice", "Bob", "Charlie"]
scores = [85, 90, 78]

combined = zip(names, scores)
combined

### **Observing Zipped Contents Using a For Loop**
* The `zip()` function stores the combined elements in tuples.
* For loop can be used to print and observe what the zipped contents look like
* The For loop output shows how the zip function combines elements from the input Lists into Tuples.

In [ ]:
for info in combined:
    print(info)

('Alice', 85)
('Bob', 90)
('Charlie', 78)


### **Important note about zip object as an iterator**
* The zip object is an iterator, which means it can only be traversed once.
* After you've iterated over it (like in the previous for loop), it gets exhausted, and you can't use it again unless you recreate it.
* This is the reason the below for loop does not print anything

In [ ]:
for info in combined:
    print(info)

### **Converting Zipped object into another Iterable**
* Notice that in the below code examples, the zip object is recreated each time before being converted to a :
 * List
 * Tuple

In [ ]:
combined = zip(names, scores)
list(combined)

[('Alice', 85), ('Bob', 90), ('Charlie', 78)]

In [ ]:
combined = zip(names, scores)
tuple(combined)

(('Alice', 85), ('Bob', 90), ('Charlie', 78))

## **Use of `zip()` function in running a For loop**

### **Basic use**

* Consider again the two lists: names and scores

In [ ]:
names = ["Alice", "Bob", "Charlie"]
scores = [85, 90, 78]

* As seen earlier, the zip() function pairs each element from the names list with the corresponding element from the scores list.
* This pairing is done element by element, creating tuples. So, ("Alice", 85), ("Bob", 90), and ("Charlie", 78) are the tuples created and stored within the zip object.

In [ ]:
combined = zip(names, scores)
for name, score in combined:
    print(f"{name}: {score}")

Alice: 85
Bob: 90
Charlie: 78


* When you use the for loop **`for name, score in combined :`** to iterate over the zip object, each tuple is unpacked into two separate variables: name and score. This is the Tuple unpacking at work
* On each iteration of the loop, name takes the first value of the tuple (e.g., "Alice"), and score takes the second value (e.g., 85).
* The loop then processes these individual values, printing them as specified.

### **Practical use of `zip()` function in running a For loop on two variables**

##### **Example 1: Computing Total Sales for Different Years**
* Suppose you have sales data for different years and you want to compute and display the total sales for each year:

In [ ]:
years = [2019, 2020, 2021]
sales = [[1200, 1500, 1800],  # Sales data for 2019
         [1300, 1600, 1900],  # Sales data for 2020
         [1400, 1700, 2000]]   # Sales data for 2021

for year, year_sales in zip(years, sales):
    total_sales = sum(year_sales)
    print(f"Total sales for {year}: {total_sales}")

Total sales for 2019: 4500
Total sales for 2020: 4800
Total sales for 2021: 5100


##### **Example 2: Computing Net profit for Products**
* Consider Three Lists `products`, `sales` and `expenses` for different products
* You want to calculate and display the Net Profit for each product

In [ ]:
products = ["Product A", "Product B", "Product C"]
sales = [1200, 1500, 1000]
expenses = [700, 800, 600]

for product, sale, expense in zip(products, sales, expenses):
    net_profit = sale - expense
    print(f"Net profit for {product}: ${net_profit}")

Net profit for Product A: $500
Net profit for Product B: $700
Net profit for Product C: $400


##### **Example 3 : Computing the Return on Investment (ROI) for each campaign**

In [ ]:
campaign_dates = ["January 2024", "February 2024", "March 2024"]
budgets = [5000, 7000, 6000]
revenues = [15000, 21000, 18000]

for date, budget, revenue in zip(campaign_dates, budgets, revenues):
    if budget > 0:
        roi = (revenue - budget) / budget
    else:
        roi = float('inf')  # Handle division by zero if budget is 0
    print(f"Campaign Date: {date} - ROI: {roi:.2f}")

Campaign Date: January 2024 - ROI: 2.00
Campaign Date: February 2024 - ROI: 2.00
Campaign Date: March 2024 - ROI: 2.00


# **Section 04 : Using For Loop with Else Statement**

### **Overview**
* In Python, you can use an `else` statement with a `for loop`.
* This might seem unusual since `else` is often used with `if` statements, but it has a specific role when used with loops.
* This section explores why and how you would use `else` with a `for loop`

### **Basic Example to Understand the working**
**Explanation for the below example**
* **Looping through Items:**
 * The for loop iterates over each `item` in the `items` list.
 * For each `item`, it checks if `item` is equal to search.
   * If the condition `item == search` is `True`, it prints a message `"Found 3!"` and uses `break` to exit the loop immediately.
   * If `break` is executed, the `else block` is skipped.

* **Else Block:**
 * The `else block` runs only if the loop completes all its iterations without encountering a `break`.
 * In this case, if search was not found in the list, the `else block` would print "3 not found.".

In [ ]:
items = [1, 2, 3, 4, 5]
search = 3

for item in items:
    if item == search:
        print(f"Found {search}!")
        break
else:
    print(f"{search} not found.")

Found 3!


**Why This Example?**
* Although checking membership using the `in` operator is more efficient for this task (i.e., when searching in a List), this example illustrates how the `else block` works with a `for loop`

* This example helps understand how the else block can be used to execute code if a loop runs to completion without a `break`
* It’s useful for scenarios where you want to take action if a loop completes all iterations and no early exit occurs

### **A more practical example to understand Application**

### **Checking if a Number is Prime**

* In this example, we want to determine if a number `num`, is a prime number.
* _A prime number is a number greater than 1 that has no positive divisors other than 1 and itself_
* To check if a number is prime, we need to verify that it isn’t divisible by any number other than 1 and itself.

In [ ]:
num = 44
for i in range(2,num):
  if num % i == 0:
    print(f"{num} is not Prime")
    break
else:
  print(f"{num} is Prime")

44 is not Prime


**Explanation**

* We start by setting `num` to 44. We want to check if 44 is a prime number.

* Looping Through Possible Divisors:

 * The for loop iterates over all numbers from 2 up to, but not including, `num` (so, 2 to 43 in this case).
 * For each number `i`, we check if it divides num without leaving a remainder (i.e., `num % i == 0`).

* Checking for Divisors:

 * If the condition `num % i == 0` is `True`, it means `i` is a divisor of `num`. Therefore, `num` is not a prime number.
 * The print statement inside the if block informs us that `num` is not prime, and the break statement exits the loop immediately.

* `Else Block`:

 * If the loop completes without finding any divisors (meaning `num` is not divisible by any number between 2 and `num-1`), the `else block` executes.
 * The `else block` prints that `num` is a prime number, as no divisors other than 1 and itself were found.

**Why use `else` Here?**

* The `else block` provides a clear way to handle the case where no divisors are found. It simplifies the code by eliminating the need for additional checks outside the loop.
* Using `else` keeps the logic within the loop and avoids extra conditional checks, making the code more readable and concise.

### **Practice Problem: Surgical Checklist Validation**
* _Feel free to practice this problem to get a better understanding of using the else statement with a for loop._

* **Scenario:** Before a critical surgery, it's crucial to ensure that all _mandatory checks_ are completed. If any check is missing, the surgery cannot proceed. Your task is to write a Python program that validates a checklist of _performed checks_ against a list of mandatory checks.

* **Objective:** Create a Python program that uses a for loop with an else block to verify that all required checks have been completed before allowing the surgery to proceed.

In [ ]:
# List of mandatory checks
mandatory_checks = [
    'Patient identity confirmed',
    'Surgical site marked',
    'Informed consent obtained',
    'Antibiotic prophylaxis administered',
    'Allergies reviewed',
    'Surgical instruments sterilized']

# Checklist of checks performed
checklist_performed = [
    'Patient identity confirmed',
    'Surgical site marked',
    'Informed consent obtained',
    'Antibiotic prophylaxis administered',
    'Allergies reviewed']

#-------------Write your code here------------

### **End of Document**